### PR TITLE
fuzzy DateTime parsing

### PR DESCRIPTION
**Pull request by robcza**
Increase the possibility of successful DateTime parsing by enabling fuzzy parser. This also allows removing explicit fuzzy parsing from the CSV parser and rely on the sanitization process only.

### CODE REVIEW

**Comment by aaronkaplan** [comment]
the automatic test is failing:

```
ERROR: Test if correct Event has been produced.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/certtools/intelmq/intelmq/tests/bots/parsers/generic/test_parser_csv.py", line 60, in test_event
    self.assertMessageEqual(0, EXAMPLE_EVENT)
  File "/home/travis/build/certtools/intelmq/intelmq/lib/test.py", line 369, in assertMessageEqual
    event = self.get_output_queue()[queue_pos]
nose.proxy.IndexError: list index out of range
```

**Comment by sebix** [comment]
The relevant part is this:
```python
test-bot: ERROR: Failed to parse line.
Traceback (most recent call last):
  File "/home/travis/build/certtools/intelmq/intelmq/lib/bot.py", line 448, in process
    events = list(filter(bool, self.parse_line(line, report)))
  File "/home/travis/build/certtools/intelmq/intelmq/bots/parsers/generic/parser_csv.py", line 90, in parse_line
    event.add(key, value)
  File "/home/travis/build/certtools/intelmq/intelmq/lib/message.py", line 121, in add
    raise exceptions.InvalidValue(key, old_value)
intelmq.lib.exceptions.InvalidValue: invalid value '2015-12-14 04:19:00' (<class 'str'>) for key 'time.source' : 
```

And further:
```python
Traceback (most recent call last):
  File "/home/travis/build/certtools/intelmq/intelmq/lib/bot.py", line 121, in start
    self.process()
  File "/home/travis/build/certtools/intelmq/intelmq/lib/bot.py", line 456, in process
    self._dump_message(exc, self.recover_line(line))
  File "/home/travis/build/certtools/intelmq/intelmq/lib/bot.py", line 314, in _dump_message
    new_dump_data[timestamp]["message"] = message.serialize()
AttributeError: 'str' object has no attribute 'serialize'
```
But for the second, it's probably me to blame.

**Comment by robcza** [comment]
Well, let's do this one thing at a time. I wanted to adjust the csv parser as well, which turns out unwise.
Adding the fuzziness to the DateTime.__parse seems however quite helpful in lot of our use cases. According to my understanding, this shouldn't break anything, just increase chances of parsing weird timestamps.

**Comment by sebix** [comment]
Looks better now :)

I don't question the reasonableness, though: do you have examples where this is useful?

**Comment by aaronkaplan** [comment]
...or one or two test cases ;) 

---
Mobile

**Comment by robcza** [comment]
Well, I've encountered situation, where the generic csv parser produces timestamps not accepted by the DateTime without fuzziness. Fixing the parser was one way, however I decided to go directly to the lib, to make the adjustment global and not to put the pressure on the bot programmer.

The problem is probably in the way I work with timezones in the generic csv parser. I append the string ' UTC' to any timestamp without knowing anything about the timezone. This breaks the sanitization without fuzziness:
```
>>> from dateutil.parser import parse
>>> import intelmq.lib.harmonization
>>> time = parse('2016-07-19 06:40:01.617719 CET +0000', fuzzy=True).isoformat()
>>> intelmq.lib.harmonization.DateTime.sanitize(time + ' UTC')
>>> 
```

However sanitization with fuzziness actually deals nicely even with the timezone and returns the expected value:
```
>>> from dateutil.parser import parse
>>> import intelmq.lib.harmonization
>>> time = parse('2016-07-19 06:40:01.617719 CET +0000', fuzzy=True).isoformat()
>>> intelmq.lib.harmonization.DateTime.sanitize(time + ' UTC')
'2016-07-19T04:40:01.617719+00:00'
>>>
```

**Comment by aaronkaplan** [comment]
sorry for being so pesky, but would you mind making this example above into a actual test case where fuzzy-ness is tested in practice?
Something like 

``
  AssertEqual(ts, '2016-07-19T04:40:01.617719+00:00')
``
?

**Comment by robcza** [comment]
@aaronkaplan it is not you being pesky, it is me being sloppy ;)
Test passing locally, however failing in Travis. I'll try to find the problem

**Comment by robcza** [comment]
Something is wrong, this is my local test output:
```
# python3 intelmq/tests/lib/test_harmonization.py 
.....................................
----------------------------------------------------------------------
Ran 37 tests in 0.071s

OK
```

Could this be somehow connected to the local time-zone setting?

**Comment by sebix** [comment]
Dependence on local timzones...

```bash
======================================================================
FAIL: Test DateTime.sanitize method.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/certtools/intelmq/intelmq/tests/lib/test_harmonization.py", line 182, in test_datetime_sanitize
    '2016-07-19 06:40:01.617719 CET +0000 UTC'))
AssertionError: '2016-07-19T04:40:01.617719+00:00' != '2016-07-19T06:40:01.617719+00:00'
- 2016-07-19T04:40:01.617719+00:00
?             ^
+ 2016-07-19T06:40:01.617719+00:00
?             ^
```
This is the first of your 2 added test cases.

File: intelmq/tests/lib/test_harmonization.py [hunk 1]
```diff
@@ -175,6 +175,15 @@ def test_datetime_from_timestamp(self):
...
                                                                'Guyana'))
 
+    def test_datetime_sanitize(self):
+        """ Test DateTime.sanitize method. """
+        self.assertEqual('2016-07-19T04:40:01.617719+00:00',
+                         harmonization.DateTime.sanitize(
+                         '2016-07-19 06:40:01.617719+02:00 UTC'))
+        self.assertEqual('2016-07-19T13:08:38+00:00',
+                         harmonization.DateTime.sanitize(
+                         '2016-07-19 13:08:38 UTC'))
```
Comment (sebix) [inline intelmq/tests/lib/test_harmonization.py]: This test also works without `fuzzy=True`.

Comment (robcza) [inline intelmq/tests/lib/test_harmonization.py]: It does not in my case
```
# python3 intelmq/tests/lib/test_harmonization.py 
.......F.............................
======================================================================
FAIL: test_datetime_sanitize (__main__.TestHarmonization)
Test DateTime.sanitize method.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "intelmq/tests/lib/test_harmonization.py", line 182, in test_datetime_sanitize
    '2016-07-19 06:40:01.617719+02:00 UTC'))
AssertionError: '2016-07-19T04:40:01.617719+00:00' != None

----------------------------------------------------------------------
Ran 37 tests in 0.044s

FAILED (failures=1)
```

Comment (sebix) [inline intelmq/tests/lib/test_harmonization.py]: I'm talking about the second one.

Comment (robcza) [inline intelmq/tests/lib/test_harmonization.py]: That is right, I've added another sample just to have more things covered. Did not mean the test as fuzzy only.

**Comment by aaronkaplan** [comment]
confirmed with @robcza that this works for him now.

**Pull request merged**